src/libgap-api.c

```diff
@@ -337,3 +341,69 @@ Obj GAP_CharWithValue(UChar obj)
 {
     return ObjsChar[obj];
 }
+
+syJmp_buf * GAP_GetReadJmpError(void)
+{
+    return &(STATE(ReadJmpError));
+}
+
+
+static volatile sig_atomic_t EnterStackCount = 0;
```

Comment:
None specifically, but my concern here is that one could: E.g. a signal handler installed by GAP which runs and results in an error, which can in turn result in running code that does modify this variable.  I could try to concoct a real example, but something like that seems plausible.

Reply:
GAP never installs such a signal handler, and indeed, calling any kind of GAP code from a signal handler would be very dangerous and is unlikely to work.

The reason I am asking is not that I want to pain the bikeshed; but rather it is that using `sig_atomic_t` IMHO adds considerable to the complexity of this code. If I or somebody else need to edit that code, I may wonder "but why is it using sig_atomic_t?", and not knowing the answer means that I don't understand the code, and hence better not touch it; or touch it but with a bad feeling that I might be breaking something.

This is why I ask whether it is really necessary, and if so, that a comment be added explaining it. Even if that comment says something like "// It is unclear whether we need sig_atomic_t here, but just out of paranoia, we use it anyway".

That said, I'd still prefer if it was not used; I honestly can't think of any example where it might be used that is not horribly unsafe anyway. So if you can think of any, I'd be most interested to hear about it, even if it is just a rough sketch.

The argument that one "could" need it seems week to me: with that argument, perhaps we should also add a mutex protecting access to `EnterStackCount` (or require C11 atomics, or something like that)? After all, multiple threads might want to use the GAP API. Sure, it's unlikely that this can work, but one could try it anyway.... ?

Reply:
I'm not sure on what basis you can assert that it "adds considerable...complexity".  Exactly what complexity is it adding?  `sig_atomic_t` is standard POSIX, and all it means is that it's an instruction to the compiler to only manipulate this variable with single instructions.  Furthermore, according to the [GCC docs](https://www.gnu.org/software/libc/manual/html_node/Atomic-Types.html), "In practice, you can assume that `int` is atomic...these assumptions are true on all of the machines that the GNU C Library supports and on all POSIX systems we know of."

It's really just a guarantee that if this code happens to be interrupted by a signal (with a signal handler that returns) it will leave things in a sane state.  For a signal handler that doesn't return--i.e. exits the process or longjumps to somewhere arbitrary) you can't make any further guarantees, but at the very least you can safely increment and decrement it even if a signal occurs. That's all.

Reply:
> at the very least you can safely increment and decrement it even if a signal occurs. That's all.

Just to clarify this: the type `sig_atomic_t` means that it can be read from atomically and written to atomically. Any operation more complicated than that (such as an increment/decrement) may not be atomic. For example, the operations `EnterStackCount = -EnterStackCount;` and `EnterStackCount++;` may not be atomic.

So I actually agree with @fingolfin here: better not make any guarantees about atomicity.

Reply:
Yep, you guys are right, and I was wrong.  My apologies. 

For Sage, Jeroen suggested that we use the functions from cysignals `sig_block()` and `sig_unblock()` around `GAP_Enter()` and `GAP_Leave()`.  The effect of this (and Jeroen can correct me if this isn't quite right) is that if an interrupt-like signal occurs during `sig_block()` (e.g. `SIGINT`, `SIGALRM` or `SIGHUP`) it quickly exits the signal handler and defers handling of the signal until later (e.g. when the Python interpreter resumes).

That's fine for Sage, but in general it doesn't make sense for this interface to have anything to do with cysignals.  I wonder if it *should* try to do anything at all.  For example, perhaps `GAP_Enter()` and `GAP_Leave()` could use `sigprocmask` to momentarily block interrupts (just within the macros).  Would that make any sense?  Jeroen has thought about these things much more deeply than I have.

Reply:
>  For example, perhaps `GAP_Enter()` and `GAP_Leave()` could use `sigprocmask` to momentarily block interrupts (just within the macros).

Since `sigprocmask()` is a system call, that would have significant impact on performance. For simple GAP calls, it would certainly dominate the runtime. On my GNU/Linux laptop, it takes about 100 clock cycles for one such call. On other OSes, it can be much worse.

In cysignals, I make a lot of effort to avoid syscalls on the critical path.

Reply:
I don't fully understand what `GAP_Enter` and `GAP_Leave` do precisely, but it might be possible to make small adjustments to the code to make it safer. For example, saving the value of `EnterStackCount` to a temporary variable in `GAP_Enter` and then restoring it from that variable in `GAP_Leave` would work. It would require a bit of C macro hacking to get it to work, but I think that it could be done.

Reply:
True--in that case it's probably best avoided. I know in Cysignals itself `sig_block()/unblock()` just set a flag that determines what to do in the actual signal handler should a signal occur.  But in the case of libgap we want to leave most details of signal handling up to the embedding application, so it would be tricky to implement something like this in a way that can be sensibly embedded in arbitrary applications.  So perhaps it's best to just keep it simple.

Regarding your idea, I'm not saying it can't be done. I just don't understand exactly what you're proposing.  Keep in mind these macros should be able to work properly recursively.  Though the exact value of EnterStackCount is less important than just tracking whether or not we're in a recursive call.

Reply:
Obvious question: do we really need to support nesting for these macros? It would certainly simplify things if we could say that nesting is not allowed.

Reply:
In theory one could write code such that they never have to be nested.  For example there are cases where we could refactor a function `some_func_that_uses_gap()` into  `some_func_that_uses_gap_outer()` and `some_func_that_uses_gap_inner(), where the latter can be called directly in other code that is already bracketed by the Enter/Leave macros.

So technically it's possible.  But it does put a higher strain on the user in exactly how they structure their code, and it's ugly (imagine if sig_on and sig_off could not be nested).  You would also have to actively prevent the user from nesting them somehow if it did not work (otherwise this would lead to hard to trace bugs).  I think it's better to allow nesting.  Any complication that arises from that seems better and less probable than not allowing it.

Reply:
For now, I propose a simple and concrete solution:

- In GAP, don't try to make these macros signal-safe. Do you know about other applications using libGAP which also need to handle signals?
- In SageMath, wrap `GAP_Enter`/`GAP_Leave` in `sig_block`/`sig_unblock` calls.

Reply:
Which is essentially the status quo (we don't have the Sage part yet but that's easy, and I agree with it).  So I'm fine with that for lack of a better solution.

Perhaps at most this should be documented, and left to individual users to worry about.  E.g. for the GAP-Julia interface something might want to be done, but it would be just as Julia-specific as Cysignals is Python-specific.  Other applications may not care as much about handling signals at all.